### ISSUE TITLE
Cache won't be cleared properly when some cache elements are removed manually

### ISSUE THREAD

**Issue by mateuszbieniek** [symfony/symfony#43823]
### Symfony version(s) affected

5.3.10

### Description

On deployment, we use
```
rm -Rf var/cache/${APP_ENV-dev}/*.*
```
command to avoid issues on interface changes. Afterward, standard
```
php bin/console cache:clear
```
is called.

This results in the cache being cleared only partially (for example template and translation cache are not cleated).

### How to reproduce

1. Clean installation of Sf 5.3.10 with `APP_ENV` set to prod
2. Execute 
```
php bin/console c:c
```
3. List files in the `var/cache/prod` directory, it will look something like (notice the dates)
```
-rw-rw-r--  1 bulbo bulbo  3142 paź 29 11:46 annotations.map
-rw-rw-r--  1 bulbo bulbo 27628 paź 29 11:46 annotations.php
-rw-rw-r--  1 bulbo bulbo   743 paź 29 11:46 App_KernelProdContainer.php
-rw-rw-r--  1 bulbo bulbo     0 paź 29 11:46 App_KernelProdContainer.php.lock
-rw-rw-r--  1 bulbo bulbo 15795 paź 29 11:46 App_KernelProdContainer.php.meta
-rw-rw-r--  1 bulbo bulbo 33308 paź 29 11:46 App_KernelProdContainer.preload.php
drwxrwxr-x  2 bulbo bulbo 20480 paź 29 11:46 ContainerSbyAQjn
drwxrwxr-x  3 bulbo bulbo  4096 paź 29 11:45 doctrine
drwxrwxr-x 11 bulbo bulbo  4096 paź 29 11:46 pools
-rw-rw-r--  1 bulbo bulbo   100 paź 29 11:46 serialization.php
drwxrwxr-x  3 bulbo bulbo  4096 paź 29 11:45 Symfony
drwxrwxr-x  2 bulbo bulbo  4096 paź 29 11:45 translations
drwxrwxr-x 25 bulbo bulbo  4096 paź 29 11:45 twig
-rw-rw-r--  1 bulbo bulbo   217 paź 29 11:46 url_generating_routes.php
-rw-rw-r--  1 bulbo bulbo  1768 paź 29 11:46 url_generating_routes.php.meta
-rw-rw-r--  1 bulbo bulbo   399 paź 29 11:46 url_matching_routes.php
-rw-rw-r--  1 bulbo bulbo  1768 paź 29 11:46 url_matching_routes.php.meta
-rw-rw-r--  1 bulbo bulbo  5968 paź 29 11:46 validation.php
```
4. Remove `*.*` files from the cache directory
```
rm -rf var/cache/prod/*.*
```
5. Clear the cache 
```
php bin/console c:c
```
4. List files in the `var/cache/prod` again
```
-rw-rw-r--  1 bulbo bulbo  3142 paź 29 12:02 annotations.map
-rw-rw-r--  1 bulbo bulbo 27628 paź 29 12:02 annotations.php
-rw-rw-r--  1 bulbo bulbo   743 paź 29 12:02 App_KernelProdContainer.php
-rw-rw-r--  1 bulbo bulbo     0 paź 29 12:02 App_KernelProdContainer.php.lock
-rw-rw-r--  1 bulbo bulbo 15795 paź 29 12:02 App_KernelProdContainer.php.meta
-rw-rw-r--  1 bulbo bulbo 33308 paź 29 12:02 App_KernelProdContainer.preload.php
drwxrwxr-x  2 bulbo bulbo 20480 paź 29 12:02 ContainerSbyAQjn
drwxrwxr-x  3 bulbo bulbo  4096 paź 29 11:45 doctrine
drwxrwxr-x 11 bulbo bulbo  4096 paź 29 11:46 pools
-rw-rw-r--  1 bulbo bulbo   100 paź 29 12:02 serialization.php
drwxrwxr-x  3 bulbo bulbo  4096 paź 29 11:45 Symfony
drwxrwxr-x  2 bulbo bulbo  4096 paź 29 11:45 translations
drwxrwxr-x 25 bulbo bulbo  4096 paź 29 11:45 twig
-rw-rw-r--  1 bulbo bulbo   217 paź 29 12:02 url_generating_routes.php
-rw-rw-r--  1 bulbo bulbo  1768 paź 29 12:02 url_generating_routes.php.meta
-rw-rw-r--  1 bulbo bulbo   399 paź 29 12:02 url_matching_routes.php
-rw-rw-r--  1 bulbo bulbo  1768 paź 29 12:02 url_matching_routes.php.meta
-rw-rw-r--  1 bulbo bulbo  5968 paź 29 12:02 validation.php
```

You can see that cache wasn't cleared properly.

**Comment by dmaicher** [comment]
I guess `rm -rf var/cache/prod/*.*` only deletes files directly inside the `cache/prod` folder :wink: (or actually everthing that matches the naming `*.*` with a dot)

Maybe try `rm -rf var/cache/prod/*` to actually purge the folder completely?

**Comment by crevillo** [comment]
Also worth to mention two more things
1) If you just do `php bin/console c:c` cache is cleared as expected (template and translation)
2) if you do `php bin/console c:c && rm -Rf var/cache/prod/*.*` then cache is cleared as expected as well. (template and translation) 

So the issue is when the files are removed before trying the cache:clear.

**Comment by crevillo** [comment]
Looks somehow related to how `CacheClearCommand` 
```
 if ($_SERVER['REQUEST_TIME'] <= filemtime($containerFile) && filemtime($containerFile) <= time()) {
 // cache is fresh 
} else {
//
}
```

When removing the files with rm -rf `var/cache/prod/*.*`, the next execution of `bin console c:c` will create the`containerFile` again. The condition above in that case evaluates to true. If you do that again, as the container file was already there, the filemtime for the container file is less than the _$_SERVER['REQUEST_TIME'] and it evaluates to false. 

Maybe because of this the translation folder and it's contents are not being regenerated (In fact i'm noticing that when cache is fresh no translation cache is being written at all...)

**Comment by crevillo** [comment]
Ok, some more findings. I don't think this is a Symfony bug. 

As explained above, when we remove the container file and do a cache clear afterwards, the command goes for the "cache is fresh" part and it does not perform the steps to create that *warmup* directory. Instead it tries to evaluate if a file like `var/cache/prod/translations/catalogue.en.g7Xb65G.php` is present as is not fresh. 

We can "hack" the code at https://github.com/symfony/config/blob/5.3/ResourceCheckerConfigCacheFactory.php#L35-L42
and leave as 
``` php
public function cache(string $file, callable $callable)
    {
        $cache = new ResourceCheckerConfigCache($file, $this->resourceCheckers);
        if (!$cache->isFresh()) {
            print $file . "\n";
            $callable($cache);
        }

        return $cache;
    }
```

Doing like that, if do a `bin/console cache:clear -v` I can see 

```
www-data@28324ccfff56:/app$ php bin/console c:c -v

 // Clearing the cache for the prod environment with debug false                                                        

 // Warming up cache...                                                                                                 

/app/var/cache/pro_/translations/catalogue.en.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.af.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ar.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.az.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.be.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.bg.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.bs.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ca.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.cs.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.cy.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.da.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.de.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.el.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.es.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.et.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.eu.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fa.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fi.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.gl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.he.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hu.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hy.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.id.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.it.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ja.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lb.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lt.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lv.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.mn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.my.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nb.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.no.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pt.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pt_BR.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ro.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ru.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sk.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sq.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sr_Cyrl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sr_Latn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sv.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.th.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.tl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.tr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.uk.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.uz.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.vi.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.zh_CN.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.zh_TW.g7Xb65G.php
/app/var/cache/pro_/url_matching_routes.php
/app/var/cache/pro_/url_generating_routes.php
 // Removing old build and cache directory...                                                                           

 // Finished
```

but if do `rm -Rf var/cache/prod/*.* && bin/console cache:clear -v`

```
www-data@28324ccfff56:/app$ rm -Rf var/cache/prod/*.* && bin/console c:c -v

 // Clearing the cache for the prod environment with debug false                                                        

 // Cache is fresh.                                                                                                     

 // Warming up optional cache...                                                                                        

/app/var/cache/prod/url_matching_routes.php
/app/var/cache/prod/url_generating_routes.php
 // Finished  
```

Otoh if I do `rm -Rf var/cache/prod/translations/*.* && bin/console c:c -v` I can see

``` 
www-data@28324ccfff56:/app$ rm -Rf var/cache/prod/translations/*.* && bin/console c:c -v

 // Clearing the cache for the prod environment with debug false                                                        

 // Warming up cache...                                                                                                 

/app/var/cache/pro_/translations/catalogue.en.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.af.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ar.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.az.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.be.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.bg.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.bs.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ca.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.cs.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.cy.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.da.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.de.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.el.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.es.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.et.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.eu.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fa.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fi.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.fr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.gl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.he.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hu.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.hy.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.id.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.it.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ja.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lb.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lt.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.lv.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.mn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.my.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nb.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.nn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.no.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pt.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.pt_BR.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ro.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.ru.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sk.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sq.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sr_Cyrl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sr_Latn.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.sv.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.th.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.tl.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.tr.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.uk.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.uz.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.vi.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.zh_CN.g7Xb65G.php
/app/var/cache/pro_/translations/catalogue.zh_TW.g7Xb65G.php
/app/var/cache/pro_/url_matching_routes.php
/app/var/cache/pro_/url_generating_routes.php
 // Removing old build and cache directory...                                                                           

 // Finished   
```
This makes me thing that is not a symfony bug and that we may need to remove all those translations files if we want them up to date after deploy them. 

In other words, if we do `var/cache/prod/*.*` we should add `var/cache/prod/translations/*.*` (and probably others) or just `rm -rf var/cache/prod/*` as suggested by @dmaicher 

 Given this is platform.sh and we do this on deploy step, I don't see a big problem of removing all files. Obviously the warmup may take a bit more time, but still the site is not up at that time, right?

**Comment by mateuszbieniek** [comment]
For simple debugging, I've changed this part
```
        if ($_SERVER['REQUEST_TIME'] <= filemtime($containerFile) && filemtime($containerFile) <= time()) {
```
to
```
        echo '$_SERVER[\'REQUEST_TIME\']: ';
        dump($_SERVER['REQUEST_TIME']);
        echo 'filemtime($containerFile): ';
        dump(filemtime($containerFile));
        echo 'time(): ';
        dump(time());
        die();
        if ($_SERVER['REQUEST_TIME'] <= filemtime($containerFile) && filemtime($containerFile) <= time()) {
```

Now when `cache:clear` command is executed when the cache is present inside `var/cache/prod` we can see everything looks normal:
```
php bin/console c:c --env=prod

 // Clearing the cache for the prod environment with debug                      
 // false                                                                       

$_SERVER['REQUEST_TIME']: ^ 1635942500
filemtime($containerFile): ^ 1635942488
time(): ^ 1635942500
```

But when `rm -rf var/cache/prod/*.*` is executed beforehand, the script is "tricked" into thinking that cache is fresh (and it isn't):
```
bulbo@bulboPC:~/www/experience-3.3$ rm -rf var/cache/prod/*.*
bulbo@bulboPC:~/www/experience-3.3$ php bin/console c:c --env=prod

 // Clearing the cache for the prod environment with debug                      
 // false                                                                       

$_SERVER['REQUEST_TIME']: ^ 1635942532
filemtime($containerFile): ^ 1635942536
time(): ^ 1635942536
```

You can notice, that `filemtime($containerFile)` equals to `time()` and is bigger than `$_SERVER['REQUEST_TIME']`

This happens because the container directory will be modified AFTER the script will start (regenerating files missing from `rm -rf`). So the issue here is that (probably) this part can be improved to avoid such "false positives".

**Comment by nicolas-grekas** [comment]
You might want to run `rm -rf var/cache/prod/*` instead.
I'm closing because there is no bug in Symfony here.
